### DISCUSSION TITLE
Recommended approach to test API rules and filters?

### DISCUSSION THREAD

**Discussion by kisamoto** [pocketbase/pocketbase#4424] [Q&A]
To ensure I have configured my API rules & filters correctly I would like to write tests that mimic requests from the client SDK and check behaviour e.g. users cannot access other users data.

I am extending with Go and have read the docs on [testing custom endpoints](https://pocketbase.io/docs/go-testing/) however as this is a non-custom, "generated" endpoint I wanted to check if there is a recommended approach to achieving this.

## Option 1: Follow same pattern as Go but use the builtin web API URLs

This would essentially mimic the process in the testing custom endpoints docs however we would replace the custom endpoint with the [Web API](https://pocketbase.io/docs/api-records#create-record) endpoints.

## Option 2: Write E2E tests in client code

Rather than embedding the tests in Go, this approach would be to set up a testing environment in my Javascript code that uses the JS SDK to make requests and checks the response. 

Any thoughts or recommendations for best practices about how to approach this?

Thanks & happy hacking

**Comment by ganigeorgiev** [comment]
Personally I use "Option 1" in my other project - https://github.com/presentator/presentator.

"Option 2" in my opinion would be more suitable for the cases when you are not extending PocketBase with Go. It is a valid approach but requires creating and configuring manually a test environment that also will have to make sure that the PocketBase process is running and that the test data is properly reset on each test run.

**Reply by kisamoto** [reply]
Thanks for the swift reply. I will check out the Presentator code base for any examples and will continue to implement my tests that way.

All the best
